bookkeeper-common/src/main/java/org/apache/bookkeeper/common/conf/ConfigDef.java

```diff
@@ -0,0 +1,325 @@
+            checkArgument(
+                Sets.difference(
+                    groups.stream().map(group -> group.name()).collect(Collectors.toSet()),
+                    settings.keySet()
+                ).isEmpty(),
+                "Configuration Key Groups doesn't match with keys");
+            return new ConfigDef(groups, settings);
+        }
+
+    }
+
+    /**
+     * Create a builder to build a config def.
+     *
+     * @return builder to build a config def.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    private final Set<ConfigKeyGroup> groups;
+    private final Map<String, Set<ConfigKey>> settings;
+    private final Map<String, ConfigKey> keys;
+
+    private ConfigDef(Set<ConfigKeyGroup> groups,
+                      Map<String, Set<ConfigKey>> settings) {
+        this.groups = groups;
+        this.settings = settings;
+        this.keys = settings.values()
+            .stream()
+            .flatMap(keys -> keys.stream())
+            .collect(Collectors.toSet())
+            .stream()
+            .collect(Collectors.toMap(
+                key -> key.name(),
+                key -> key
+            ));
+    }
+
+    /**
+     * Validate if the provided <tt>conf</tt> is a valid configuration of this configuration definition.
+     *
+     * @param conf the configuration to validate
+     */
+    public void validate(Configuration conf) throws ConfigException {
+        for (ConfigKey key : keys.values()) {
+            key.validate(conf);
+        }
+    }
+
+    /**
+     * Build the config definitation of a config class.
+     *
+     * @param configClass config class
+     * @return config definition.
+     */
+    @SuppressWarnings("unchecked")
+    public static ConfigDef of(Class configClass) {
+        ConfigDef.Builder builder = ConfigDef.builder();
+
+        Field[] fields = configClass.getDeclaredFields();
+        for (Field field : fields) {
+            if (Modifier.isStatic(field.getModifiers()) && field.getType().equals(ConfigKey.class)) {
+                field.setAccessible(true);
+                try {
+                    builder.withConfigKey((ConfigKey) field.get(null));
+                } catch (IllegalAccessException e) {
+                    log.error("Illegal to access {}#{}", configClass.getSimpleName(), field.getName(), e);
+                }
+            }
+        }
+
+        return builder.build();
+    }
+
+    //
+    // Methods to save the configuration to an {@link OutputStream}
+    //
+
+    private static final int MAX_COLUMN_SIZE = 80;
+    private static final String COMMENT_PREFIX = "# ";
+
+    public void save(Path path) throws IOException  {
+        try (OutputStream stream = Files.newOutputStream(
+            path, StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING)) {
+            save(stream);
+        }
+    }
+
+    public void save(OutputStream os) {
+        try (PrintStream ps = new PrintStream(os)) {
```

Comment:
Is this Platform dependant? Thinking about EOL

Reply:
EOL should be already handled. what is your concern here?

Reply:
If you are saving this file on linux you wil use LF, on Mac CR...
Maybe this is not a real problem, as the original file will already be using such EOL convention

Reply:
that has to be platform dependent. so I don't think that is an issue.